Sources/SourceKitLSP/Swift/SyntaxHighlightingToken.swift

```diff
@@ -0,0 +1,413 @@
+    }
+  }
+}
+
+extension Array where Element == SyntaxHighlightingToken {
+  /// The LSP representation of syntax highlighting tokens. Note that this
+  /// requires the tokens in this array to be sorted.
+  public var lspEncoded: [UInt32] {
+    var previous = Position(line: 0, utf16index: 0)
+    var rawTokens: [UInt32] = []
+    rawTokens.reserveCapacity(count * 5)
+
+    for token in self {
+      let lineDelta = token.start.line - previous.line
+      let charDelta = token.start.utf16index - (
+        // The character delta is relative to the previous token's start
+        // only if the token is on the previous token's line.
+        previous.line == token.start.line ? previous.utf16index : 0
+      )
+      previous = token.start
+      rawTokens += [
+        UInt32(lineDelta),
+        UInt32(charDelta),
+        UInt32(token.length),
+        token.kind.rawValue,
+        token.modifiers.rawValue
+      ]
+    }
+
+    return rawTokens
+  }
+
+  /// Merges the tokens in this array into a new token array,
+  /// preferring the given array's tokens if duplicate ranges are
+  /// found.
+  public func mergingTokens(with other: [SyntaxHighlightingToken]) -> [SyntaxHighlightingToken] {
+    let otherRanges = Set(other.map(\.sameLineRange))
+    return filter { !otherRanges.contains($0.sameLineRange) } + other
+  }
+}
+
+/// Parses tokens from sourcekitd response dictionaries.
+struct SyntaxHighlightingTokenParser {
+  private let sourcekitd: SourceKitD
+  /// Use `keys.nameoffset` and `keys.namelength` instead of `keys.offset`
+  /// and `keys.length`. This is useful in AST structures provided by
+  /// sourcekitd, since these contain both a name range and a normal range,
+  /// where the latter refers to the entire node (e.g. the entire function
+  /// declaration).
+  private let useName: Bool
+
+  init(sourcekitd: SourceKitD, useName: Bool = false) {
+    self.sourcekitd = sourcekitd
+    self.useName = useName
+  }
+
+  func parseTokens(_ response: SKDResponseDictionary, in snapshot: DocumentSnapshot) -> [SyntaxHighlightingToken] {
+    let keys = sourcekitd.keys
+    var tokens: [SyntaxHighlightingToken] = []
+
+    if let offset: Int = useName ? response[keys.nameoffset] : response[keys.offset],
+       var length: Int = useName ? response[keys.namelength] : response[keys.length],
+       let start: Position = snapshot.positionOf(utf8Offset: offset),
+       let skKind: sourcekitd_uid_t = response[keys.kind],
+       let (kind, modifiers) = parseKindAndModifiers(skKind) {
+
+      // We treat function declaration and enum member name tokens as a special
+      // case, e.g. SourceKit returns `f(x: Int, y: Int)` as a name instead of just `f`.
+      if useName && [.function, .method, .enumMember].contains(kind) && modifiers.contains(.declaration),
+         let name: String = response[keys.name],
+         name.contains("("),
+         let funcNameLength: Int = name.split(separator: "(").first?.count {
```

Comment:
I think this needs to be `name.split(separator: "(").first?.utf8.count`. `count` returns the number of grapheme clusters in the string, but we are counting the length as number of UTF-8 code points.

It would probably be worth adding a test case that has an emoji or something like that as a function name.

Reply:
Since LSP generally deals with positions in terms of UTF-16 code units (not code points, IIUC), I think it has to be

```swift
name.split(separator: "(").first?.utf16.count
```

But good catch, thanks!

Reply:
Actually, I think the way we are handling the lengths currently is slightly inconsistent:

- SourceKit gives us the length in UTF-8 code units (I think)
- `splitToSingleLineTokens` computes the length in terms of grapheme cluster count
- LSP expects the length in UTF-16 code units

Not sure what the ideal internal representation of `SyntaxHighlightingToken` would be (lengths in UTF-8 or UTF-16 code units). Or perhaps just using an `end: Position`, possibly requiring access to snapshots during en-/decoding as discussed [here](https://github.com/apple/sourcekit-lsp/pull/414#discussion_r671700772)?

Reply:
I think the cleanest solution would probably be to have `SyntaxHighlightingToken` have a `start` and `end` position. Together with a document snapshot, we can compute the length in any encoding we want. That gives us most flexibility and I think it the best abstraction. Good catch finding this!

Reply:
I have tried implementing this and am concerned that it complicates especially testing a fair bit (see [here](https://github.com/apple/sourcekit-lsp/pull/414#discussion_r671704692)), since we would move away from LSP's representation of tokens.

Reply:
Updated as per our discussion by using `Range<Position>` internally and asserting that tokens are on a single line.